File: NC/minimum-window-with-characters.cpp
class Solution {
public:

    bool isValidSubstring(unordered_map<char, int> &tCount,unordered_map<char, int> &sCount){
        for(auto i:tCount){
            if(i.second>sCount[i.first]){
                return false;
            }
        }
        return true;
    }

    string minWindow(string s, string t) {
        unordered_map<char, int> tCount,sCount;

        for(auto c:t){
            tCount[c]++;
        }
         
        string ans = s;
        if(s.size()<1){
            return "";
        }
        sCount[s[0]]++;

        if(isValidSubstring(tCount, sCount)){
            return s.substr(0,1);
        }

        for(int left=0, right=1;right<s.size();right++){
            char rightChar = s[right];
            sCount[rightChar]++;
            if(isValidSubstring(tCount, sCount)){
                if(right-left+1<ans.size()){
                    ans = s.substr(left, right-left+1);
                }
                while(1){
                    char leftChar = s[left];
                    sCount[leftChar]--;
                    if(isValidSubstring(tCount, sCount)){
                        left++;
                        if(right-left+1<ans.size()){
                            ans = s.substr(left, right-left+1);
                        }
                    }else{
                        sCount[leftChar]++;
                        break;
                    }
                }
            }
        }

        if(ans == s && !isValidSubstring(tCount, sCount)){
            return "";
        }
        return ans;

    }
};

File: NC/java/MinimumWindowWithCharacters.java
package NC.java;

import java.util.HashMap;
import java.util.Map;

public class MinimumWindowWithCharacters {
    private boolean isValidSubString(Map<Character, Integer> sMap, Map<Character, Integer> tMap) {
        for (var entry : tMap.entrySet()) {
            if (sMap.getOrDefault(entry.getKey(), 0) < entry.getValue()) {
                return false;
            }
        }

        return true;
    }

    public String minWindow(String s, String t) {
        if (s.equals("") || t.equals("")) {
            return "";
        }

        if (s.length() == 1 && s.equals(t)) {
            return s;
        }

        Map<Character, Integer> tMap = new HashMap();
        Map<Character, Integer> sMap = new HashMap();

        for (int i = 0; i < t.length(); i++) {
            char c = t.charAt(i);
            tMap.put(c, tMap.getOrDefault(c, 0) + 1);
        }

        boolean answerExists = false;

        String ans = s;

        for (int i = 0, j = 0; j < s.length(); j++) {
            char rightChar = s.charAt(j);
            sMap.put(rightChar, sMap.getOrDefault(rightChar, 0) + 1);

            if (isValidSubString(sMap, tMap)) {
                answerExists = true;
                char leftChar = s.charAt(i);
                while (sMap.get(leftChar) - 1 >= tMap.getOrDefault(leftChar, 0)) {
                    sMap.put(leftChar, sMap.get(leftChar) - 1);
                    i++;
                    leftChar = s.charAt(i);
                }
                if (ans.length() > j - i + 1) {
                    ans = s.substring(i, j + 1);
                }
                sMap.put(s.charAt(i), sMap.get(s.charAt(i)) - 1);
                i++;
            }
        }

        if (!answerExists) {
            return "";
        }

        return ans;

    }
}
